contrib/epee/include/syncobj.h

```diff
@@ -150,6 +154,127 @@ namespace epee
   };
 
 
+  struct reader_writer_lock {
+
+    reader_writer_lock() noexcept : 
+    write_mode(false),
+    read_mode(0) {};
+
+    bool start_read() noexcept {
+      if (have_write()) {
```

Comment:
Is this the behavior that we want: that we can't call a read function inside of a write function, even inside the same thread?

Reply:
Nevermind: this returns whether lock was just now acquired, not whether overall acquisition is a success

Reply:
Yes. `have_write` is for cases where you do have `write-lock` and recursively acquiring `read-lock` too.